logprep/connector/opensearch/output.py

```diff
@@ -137,15 +137,29 @@ def _bulk(self, client, actions, *args, **kwargs):
             self._handle_transport_error(error)
 
     def _parallel_bulk(self, client, actions, *args, **kwargs):
-        for success, item in helpers.parallel_bulk(
-            client,
-            actions=actions,
-            chunk_size=self._config.chunk_size,
-            queue_size=self._config.queue_size,
-            raise_on_error=True,
-            raise_on_exception=True,
-        ):
-            if not success:
-                result = item[list(item.keys())[0]]
-                if "error" in result:
-                    raise result.get("error")
+        bulk_delays = 1
+        for _ in range(self._config.max_retries + 1):
+            try:
+                for success, item in helpers.parallel_bulk(
+                    client,
+                    actions=actions,
+                    chunk_size=self._config.chunk_size,
+                    queue_size=self._config.queue_size,
+                    raise_on_error=True,
+                    raise_on_exception=True,
+                ):
+                    if not success:
+                        result = item[list(item.keys())[0]]
+                        if "error" in result:
+                            raise result.get("error")
+                return
+            except search.ConnectionError as error:
+                raise error
+            except search.exceptions.TransportError as error:
+                if self._message_exceeds_max_size_error(error):
+                    raise error
+                time.sleep(bulk_delays)
+                bulk_delays += random.randint(500, 1000) / 1000
+        raise FatalOutputError(
+            self, "Opensearch too many requests, all parallel bulk retries failed"
+        )
```

Comment:
It is one of these rare cases where a `for else loop` would be nice

```suggestion
         for _ in range(self._config.max_retries + 1):
            try:
                for success, item in helpers.parallel_bulk(
                    client,
                    actions=actions,
                    chunk_size=self._config.chunk_size,
                    queue_size=self._config.queue_size,
                    raise_on_error=True,
                    raise_on_exception=True,
                ):
                    if not success:
                        result = item[list(item.keys())[0]]
                        if "error" in result:
                            raise result.get("error")
                break
            except search.ConnectionError as error:
                raise error
            except search.exceptions.TransportError as error:
                if self._message_exceeds_max_size_error(error):
                    raise error
                time.sleep(bulk_delays)
                bulk_delays += random.randint(500, 1000) / 1000
        else:
            raise FatalOutputError(
                self, "Opensearch too many requests, all parallel bulk retries failed"
            )
```

I like the single method implementation more